app/scripts/controllers/transactions/pending-tx-tracker.js

```diff
@@ -157,70 +173,61 @@ export default class PendingTransactionTracker extends EventEmitter {
 
       return
     }
-    // *note to self* hard failure point
-    const transactionReceipt = await this.query.getTransactionReceipt(txHash)
-
-
-    // If another tx with the same nonce is mined, set as dropped.
-    const taken = await this._checkIfNonceIsTaken(txMeta)
-    let dropped
-    try {
-      // check the network if the nonce is ahead the tx
-      // and the tx has not been mined into a block
-      dropped = await this._checkIfTxWasDropped(txMeta, transactionReceipt)
-
-      // the dropped buffer is in case we ask a node for the tx
-      // that is behind the node we asked for tx count
-      // IS A SECURITY FOR HITTING NODES IN INFURA THAT COULD GO OUT
-      // OF SYNC.
-      // on the next block event it will return fire as dropped
-      if (typeof this.droppedBuffer[txHash] !== 'number') {
-        this.droppedBuffer[txHash] = 0
-      }
-
-      // 3 block count buffer
-      if (dropped && this.droppedBuffer[txHash] < 3) {
-        dropped = false
-        ++this.droppedBuffer[txHash]
-      }
-
-      if (dropped && this.droppedBuffer[txHash] === 3) {
-        // clean up
-        delete this.droppedBuffer[txHash]
-      }
-    } catch (e) {
-      log.error(e)
-    }
 
-    if (taken || dropped) {
+    if (await this._checkIfNonceIsTaken(txMeta)) {
       this.emit('tx:dropped', txId)
       return
     }
 
-    // get latest transaction status
-    if (transactionReceipt?.blockNumber) {
-      this.emit('tx:confirmed', txId, transactionReceipt)
-    } else {
-      const err = new Error('Missing transaction receipt or block number.')
+    try {
+      const transactionReceipt = await this.query.getTransactionReceipt(txHash)
+      if (transactionReceipt?.blockNumber) {
+        this.emit('tx:confirmed', txId, transactionReceipt)
+        return
+      }
+    } catch (err) {
       txMeta.warning = {
         error: err.message,
         message: 'There was a problem loading this transaction.',
       }
       this.emit('tx:warning', txMeta, err)
+      return
+    }
+
+    if (await this._checkIfTxWasDropped(txMeta)) {
+      this.emit('tx:dropped', txId)
+      return
```

Comment:
This check used to happen before we get the `transactionReceipt`. Why do we want to do it afterwards instead?

Reply:
We used to emit the events in this order: `tx:dropped`, then `tx:completed`, then `tx:warning`. This maintains the same order, but splits the `tx:dropped` into two as it is impossible for a tx to be both completed and dropped. The dropped check used to check if it was completed first, and that got moved.

(Wow this is a mess of words. Does this make any sense?)

Reply:
Haha, I think it did the trick!

- In essence, we need to get the transaction receipt to check if the transaction was confirmed or dropped.
- The transaction cannot be both confirmed and dropped
- Once we have the transaction receipt, checking if the transaction was confirmed is very cheap
- Therefore, check if the transaction was dropped after checking if it was confirmed

Reply:
That sounds right, at least as far as this second dropped condition is concerned. Note there are _two_ conditions where we can know the tx was dropped; the first is on line 177.

This second check always used the transaction receipt. Even before #7483, the receipt was retrieved in the `_checkIfTxWasDropped` function.